mlflow/tracking/client.py

```diff
@@ -616,15 +615,9 @@ def start_trace(
 
             trace_manager = InMemoryTraceManager.get_instance()
             tags = exclude_immutable_tags(tags or {})
-            if is_in_databricks_model_serving_environment():
-                # Update trace tags for trace in in-memory trace manager
-                with trace_manager.get_trace(request_id) as trace:
-                    trace.info.tags.update(tags)
-            else:
-                # Update trace tags in store and in-memory if tracking client is available
-                self._tracking_client.set_trace_tags(request_id, tags)
-                trace_info = self._tracking_client.get_trace_info(request_id)
-                trace_manager.update_trace_info(trace_info)
+            # Update trace tags for trace in in-memory trace manager
```

Comment:
_start_tracked_trace stores trace into the backend, then if we don't update here, the tags saved when starting trace does not include tags passed in, and if the trace exited unexpectedly the tags in backend store is not consistent

Reply:
@serena-ruan Yes, but so do any other data like inputs, span data, request_metadata. I don't see a clear reason to treat tags as a special case?

> The SetTraceTag API is called to log custom tags. However, we can record it at the end of the trace, similar to spans and any other metadata.

Reply:
I think we should fix others as well, ideally we should use the inputs/request_metadata passed in to create the trace, while span data exist in trace data so it should be fine to keep it in end_trace.

Reply:
@serena-ruan Ok, I'll see if it's feasible with current architecture. In practice, traces should always be closed, and if not it's more of an issue than whether or not tags are present🙍

Reply:
@serena-ruan I did some trials but it seems passing metadata to the initial `StartTrace` call is challenging. We call `StartTrace` within the span processor class's `on_start` handler, which only accepts the OTel span itself. This handler is executed when we create the span itself, so we cannot pass metadata via span attributes, too.

Extracting the API call out of the processor is not appealing because it will break the encapsulation of processor/exporter, which is thoughtful design of OpenTelemetry and has been working well to isolate logic between notebook and model serving.

One possible option is to make async `SetTraceTag` API request. There may be some delay compared to sync calling but at least it is guaranteed that the tags are recorded no matter the `EndTrace` is called.

However, I'm still wondering the issue here - whether tags are rendered at trace creation or end - is big enough to pay off such complexity. We haven't received any complaint on doing same for inputs and I hardly imagine any usage of tags until trace completion cuz they cannot even open the trace data until then.

This is two-way door decision anyway, we can add back `set_trace_tag` call any time if requested.